File: JAVA/ValidParentheses.java
/**
 * @version JAVA
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package JAVA;

import java.util.ArrayList;
import java.util.List;

public class ValidParentheses {

    public static void main(String[] args) {
        System.out.printf("The result is %b\n", isValid("[{}]"));
    }

    /**
     * Check if the order of opened/closed parentheses is valid.
     */
    public static boolean isValid(String s) {
        List<String> stck = new ArrayList<String>();

        for (int i=0; i < s.length(); i++) {
            String pos = "";
            if (stck.size() > 0) pos = stck.get( stck.size()-1 );

            if (s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[') {
                stck.add( String.valueOf( s.charAt(i) ) );
            } else if ( ("(".equals(pos) && s.charAt(i) == ')') || ("{".equals(pos) && s.charAt(i) == '}') || ("[".equals(pos) && s.charAt(i) == ']') ) {
                stck.remove( stck.size()-1 );
            } else {
                return false;
            }
        }

        return stck.size()>0 ? false : true;
    }
}

File: GO/ValidParentheses.go
/**
 * @version GO
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package GO

/**
 * CONNECTED WITH TEST CLASS
 */
func ValidParentheses(s string) {
}

/**
 * Check if the order of opened/closed parentheses is valid.
 */
func isValid(s string) bool {
	var stck []string
	var result bool

	for i := 0; i < len(s); i++ {
		var pos string
		if len(stck) > 0 {
			pos = stck[len(stck)-1]
		}

		if s[i] == '(' || s[i] == '{' || s[i] == '[' {
			stck = append(stck, string(s[i]))
		} else if (pos == "(" && s[i] == ')') || (pos == "{" && s[i] == '}') || (pos == "[" && s[i] == ']') {
			stck = stck[:len(stck)-1]
		} else {
			result = false
		}
	}

	if len(stck) > 0 {
		result = false
	} else {
		result = true
	}

	return result
}
